swarm/src/lib.rs

```diff
@@ -544,6 +543,24 @@ where
         self.confirmed_external_addr.iter()
     }
 
+    fn add_listener(&mut self, opts: ListenOpts) {
+        let addr = opts.address();
+        let listener_id = opts.listener_id();
+
+        if let Err(e) = self.transport.listen_on(listener_id, addr) {
+            self.behaviour
+                .on_swarm_event(FromSwarm::ListenerError(behaviour::ListenerError {
+                    listener_id,
+                    err: &e,
+                }));
```

Comment:
After this, we need to `return Err(e)`, otherwise we will always emit a `NewListener` event.

Reply:
This was overlooked on my side.